mdp/nodes/sfa_nodes.py

```diff
@@ -87,6 +118,10 @@ def __init__(self, input_dim=None, output_dim=None, dtype=None,
 
         # set routine for eigenproblem
         self._symeig = symeig
+        self._sfa_solver = (self._rank_deficit_solver_reg if handle_rank_deficit
+                else self._symeig)
+        self._rank_threshold = 1e-12
+        self.rank_deficit = 0
```

Comment:
Ideally, instead of four members, we would only have a single member variable here, e.g., `_eig_solve` and move any parameters needed for the individual solvers to their respective methods. In my mind this would make the `__init__` here a lot cleaner, which would help make the SFANode easier to understand.

Reply:
Then it would be impossible for an advanced user to configure them.
Also note that rank_deficit is meant as an output parameter, so an advanced user can transparently see how much rank deficit was detected.

Reply:
The parameters would still be here though, just part of the respective solvers. Putting them there should not keep an advanced user from setting them, but prevent a non-advanced user to accidentally mess them up ;)

Reply:
It would also take a way the option to see available parameters via inspection or the `dir` command. Additionally it would pollute the code of the solvers with existence checks and explicit default value assignments. Initializing all available parameters at one central place is IMO the cleanest and most transparent solution. IIRC adding and removing parameters after `__init__` is - while possible - considered bad style in Python OOP programming, but I might remember it wrong.

Reply:
Also note that `self._rank_threshold` is acommon parameter used by all rank deficit solvers, including [the SVD one](https://github.com/Stewori/mdp-toolkit/blob/e887365a00efe14307099e159b82231407fc7028/mdp/nodes/sfa_nodes.py#L366). So moving it to the solver code would lead to redundant code, which should also be avoided.

Reply:
Right, I can agree with those points. I would, however, make it more clear in the `__init__` how/where the parameters are used. I'd suggest once the code is feature complete then we take one of the upcoming Tuesdays, go through it together, and try to finalize it in person (still post it here before we hit the switch though, obviously).